File: src/python_gui/vdtrace_gui/depth_filters.py
from __future__ import annotations

from dataclasses import dataclass


def execution_mode_options() -> tuple[str, ...]:
    return ("EDGE", "TF")


def normalize_execution_mode(text: str) -> str:
    lowered = text.strip().lower()
    if lowered == "tf":
        return "TF"
    if lowered in ("", "edge"):
        return "EDGE"
    raise ValueError("执行模式只支持 EDGE 或 TF。")


def runtime_execution_mode_text(text: str) -> str:
    return normalize_execution_mode(text).lower()


@dataclass(slots=True)
class ModuleDepthRule:
    module_name: str
    ui_depth: str
    execution_mode: str = "EDGE"


def normalize_ui_depth_value(text: str) -> str:
    value = int(text.strip() or "0")
    if value < 0:
        raise ValueError("追踪层级必须是非负整数。")
    return str(value)


def ui_depth_to_runtime_text(text: str) -> str:
    value = int(normalize_ui_depth_value(text))
    if value == 0:
        return "all"
    if value == 1:
        return "single"
    return str(value - 1)


def parse_module_depth_rules(text: str) -> list[ModuleDepthRule]:
    rules: list[ModuleDepthRule] = []
    seen_modules: set[str] = set()
    for raw_token in [item.strip() for item in text.replace(";", ",").split(",")]:
        if not raw_token:
            continue
        parts = [item.strip() for item in raw_token.split(":")]
        if len(parts) not in (2, 3):
            raise ValueError("模块过滤规则格式无效，使用 模块名:层级[:模式]。")
        module_name = parts[0]
        if not module_name:
            raise ValueError("模块过滤规则缺少模块名。")
        lowered = module_name.lower()
        if lowered in seen_modules:
            raise ValueError(f"模块过滤规则重复：{module_name}")
        seen_modules.add(lowered)
        rules.append(
            ModuleDepthRule(
                module_name=module_name,
                ui_depth=normalize_ui_depth_value(parts[1]),
                execution_mode=normalize_execution_mode(parts[2] if len(parts) == 3 else "EDGE"),
            )
        )
    return rules


def serialize_module_depth_rules(rules: list[ModuleDepthRule]) -> str:
    tokens: list[str] = []
    for rule in rules:
        tokens.append(
            f"{rule.module_name}:{normalize_ui_depth_value(rule.ui_depth)}:{normalize_execution_mode(rule.execution_mode)}"
        )
    return ",".join(tokens)


def format_ui_depth_label(text: str) -> str:
    value = int(normalize_ui_depth_value(text))
    if value == 0:
        return "不限(ALL)"
    if value == 1:
        return "同层(SINGLE)"
    return f"向下{value - 1}层"


def format_rule_label(ui_depth: str, execution_mode: str) -> str:
    return f"{format_ui_depth_label(ui_depth)} / {normalize_execution_mode(execution_mode)}"


def build_depth_filter_spec(
    outside_enabled: bool,
    outside_ui_depth: str,
    outside_execution_mode: str,
    anonymous_enabled: bool,
    anonymous_ui_depth: str,
    anonymous_execution_mode: str,
    module_rules_text: str,
) -> str:
    tokens: list[str] = []
    if outside_enabled:
        tokens.append(
            f"outside={ui_depth_to_runtime_text(outside_ui_depth)}:{runtime_execution_mode_text(outside_execution_mode)}"
        )
    if anonymous_enabled:
        tokens.append(
            f"anon={ui_depth_to_runtime_text(anonymous_ui_depth)}:{runtime_execution_mode_text(anonymous_execution_mode)}"
        )
    for rule in parse_module_depth_rules(module_rules_text):
        tokens.append(
            "module="
            f"{rule.module_name}:{ui_depth_to_runtime_text(rule.ui_depth)}:{runtime_execution_mode_text(rule.execution_mode)}"
        )
    return ",".join(tokens)


def build_depth_filter_summary(
    default_ui_depth: str,
    outside_enabled: bool,
    outside_ui_depth: str,
    outside_execution_mode: str,
    anonymous_enabled: bool,
    anonymous_ui_depth: str,
    anonymous_execution_mode: str,
    module_rules_text: str,
    idle_escape_enabled: bool,
    idle_escape_threshold: str,
) -> str:
    parts = [f"默认={format_ui_depth_label(default_ui_depth)}"]
    if outside_enabled:
        parts.append(f"模块外={format_rule_label(outside_ui_depth, outside_execution_mode)}")
    if anonymous_enabled:
        parts.append(f"匿名页={format_rule_label(anonymous_ui_depth, anonymous_execution_mode)}")
    module_rules = parse_module_depth_rules(module_rules_text)
    if module_rules:
        tf_count = sum(1 for rule in module_rules if normalize_execution_mode(rule.execution_mode) == "TF")
        if tf_count == 0:
            parts.append(f"模块规则={len(module_rules)}条")
        else:
            parts.append(f"模块规则={len(module_rules)}条/{tf_count}个TF")
    parts.append(f"空转跳出={idle_escape_threshold.strip() or '32'}" if idle_escape_enabled else "空转跳出=关闭")
    return " | ".join(parts)

File: src/flutter_gui/lib/src/depth_filters.dart
class ModuleDepthRule
{
  const ModuleDepthRule({required this.moduleName, required this.uiDepth, this.executionMode = 'EDGE'});

  final String moduleName;
  final String uiDepth;
  final String executionMode;
}

String normalizeExecutionMode(String text)
{
  final lowered = text.trim().toLowerCase();
  if (lowered == 'tf')
  {
    return 'TF';
  }
  if (lowered.isEmpty || lowered == 'edge')
  {
    return 'EDGE';
  }
  throw FormatException('执行模式只支持 EDGE 或 TF。');
}

String runtimeExecutionModeText(String text)
{
  return normalizeExecutionMode(text).toLowerCase();
}

String normalizeUiDepthValue(String text)
{
  final value = int.parse(text.trim().isEmpty ? '0' : text.trim());
  if (value < 0)
  {
    throw FormatException('追踪层级必须是非负整数。');
  }
  return '$value';
}

String uiDepthToRuntimeText(String text)
{
  final value = int.parse(normalizeUiDepthValue(text));
  if (value == 0)
  {
    return 'all';
  }
  if (value == 1)
  {
    return 'single';
  }
  return '${value - 1}';
}

List<ModuleDepthRule> parseModuleDepthRules(String text)
{
  final rules = <ModuleDepthRule>[];
  final seen = <String>{};
  for (final raw in text.replaceAll(';', ',').split(','))
  {
    final token = raw.trim();
    if (token.isEmpty)
    {
      continue;
    }
    final parts = token.split(':').map((part) => part.trim()).toList();
    if (parts.length != 2 && parts.length != 3)
    {
      throw FormatException('模块过滤规则格式无效，使用 模块名:层级[:模式]。');
    }
    final moduleName = parts[0];
    if (moduleName.isEmpty)
    {
      throw FormatException('模块过滤规则缺少模块名。');
    }
    final lowered = moduleName.toLowerCase();
    if (seen.contains(lowered))
    {
      throw FormatException('模块过滤规则重复：$moduleName');
    }
    seen.add(lowered);
    rules.add(ModuleDepthRule(
      moduleName: moduleName,
      uiDepth: normalizeUiDepthValue(parts[1]),
      executionMode: normalizeExecutionMode(parts.length == 3 ? parts[2] : 'EDGE'),
    ));
  }
  return rules;
}

String serializeModuleDepthRules(List<ModuleDepthRule> rules)
{
  return rules
      .map((rule) => '${rule.moduleName}:${normalizeUiDepthValue(rule.uiDepth)}:${normalizeExecutionMode(rule.executionMode)}')
      .join(',');
}

String formatUiDepthLabel(String text)
{
  final value = int.parse(normalizeUiDepthValue(text));
  if (value == 0)
  {
    return '不限(ALL)';
  }
  if (value == 1)
  {
    return '同层(SINGLE)';
  }
  return '向下${value - 1}层';
}

String formatRuleLabel(String uiDepth, String executionMode)
{
  return '${formatUiDepthLabel(uiDepth)} / ${normalizeExecutionMode(executionMode)}';
}

String buildDepthFilterSpec(
  bool outsideEnabled,
  String outsideUiDepth,
  String outsideExecutionMode,
  bool anonymousEnabled,
  String anonymousUiDepth,
  String anonymousExecutionMode,
  String moduleRulesText,
)
{
  final tokens = <String>[];
  if (outsideEnabled)
  {
    tokens.add('outside=${uiDepthToRuntimeText(outsideUiDepth)}:${runtimeExecutionModeText(outsideExecutionMode)}');
  }
  if (anonymousEnabled)
  {
    tokens.add('anon=${uiDepthToRuntimeText(anonymousUiDepth)}:${runtimeExecutionModeText(anonymousExecutionMode)}');
  }
  for (final rule in parseModuleDepthRules(moduleRulesText))
  {
    tokens.add('module=${rule.moduleName}:${uiDepthToRuntimeText(rule.uiDepth)}:${runtimeExecutionModeText(rule.executionMode)}');
  }
  return tokens.join(',');
}

String buildDepthFilterSummary(
  String defaultUiDepth,
  bool outsideEnabled,
  String outsideUiDepth,
  String outsideExecutionMode,
  bool anonymousEnabled,
  String anonymousUiDepth,
  String anonymousExecutionMode,
  String moduleRulesText,
  bool idleEscapeEnabled,
  String idleEscapeThreshold,
)
{
  final parts = <String>['默认=${formatUiDepthLabel(defaultUiDepth)}'];
  if (outsideEnabled)
  {
    parts.add('模块外=${formatRuleLabel(outsideUiDepth, outsideExecutionMode)}');
  }
  if (anonymousEnabled)
  {
    parts.add('匿名页=${formatRuleLabel(anonymousUiDepth, anonymousExecutionMode)}');
  }
  final moduleRules = parseModuleDepthRules(moduleRulesText);
  if (moduleRules.isNotEmpty)
  {
    final tfCount = moduleRules.where((rule) => normalizeExecutionMode(rule.executionMode) == 'TF').length;
    parts.add(tfCount == 0 ? '模块规则=${moduleRules.length}条' : '模块规则=${moduleRules.length}条/$tfCount个TF');
  }
  parts.add(idleEscapeEnabled ? '空转跳出=${idleEscapeThreshold.trim().isEmpty ? '32' : idleEscapeThreshold.trim()}' : '空转跳出=关闭');
  return parts.join(' | ');
}
